### DISCUSSION TITLE
Web Interface

### DISCUSSION THREAD

**Discussion by Festivejelly** [kachurovskiy/nanoels#135] [Ideas]
It would be good if the ELS board could provide some sort of API so that we can build a front end for it. Right now im in building a front end for the gcode sender, it would be super handy to be able to use the functions of the controller through my ipad via a custom built UI.

I do like the simplicity of the controls on the ELS Nano but sometimes it can take quite a lot of clicks to get to the setting you want. So imagine being able to create your own drag and drop  "dashboard" so to speak that would have all the bits and bobs you need.

To be clear im not asking or a full frontend just an API that I can hook into. Id be quite happy to build and share the UI side of things when im done.

**Comment by Festivejelly** [comment]
Im thinking these commands could be sent over USB. So for example you would have a raspberry pi or PC or whatever plugged into the nano and the commands would be sent that way instead of requiring an HTTP server on the nano.

For example threading we would essentially input the same parameter as you would on the physical display, but this time this would be on a modal form in the UI, then when we click execute on the UI it would send a bunch of parameters over the USB and then that would execute on the Nano. Id image we would need some custom functions set up on the nano to listen for these various function calls.

**Reply by kachurovskiy** [reply]
Technically you could plug an antenna (neither included nor officially supported by me yet) into the ESP32 and use the WiFi capabilities on core 0. Then you could serve the Web UI directly from H4.

I'd be hesitant to offer this to users myself since WiFi is not very suitable for real-time communication and overall this seems like something that adds a lot of lines and complexity but would be used by a minority of the users. Touch screens also don't have tactile feedback which I'm a fan of :)

You'd probably want to start your own GitHub repository (fork) to develop that. Would be really cool if you could make it work.

Otherwise there's also an option to support more commands in `MODE_GCODE` but [currently we're only listening to Serial input when in that mode](https://github.com/kachurovskiy/nanoels/blob/main/h4/h4.ino#L1338) - it's probably not a big deal to listen all the time though. It would be relatively easy to receive "virtual button events" via `Serial` and enqueue them for processing by [`processKeypadEvent()`](https://github.com/kachurovskiy/nanoels/blob/main/h4/h4.ino#L2261C6-L2261C26). Perhaps there's no need for these changes to immediately go into the main version - you could keep them in your local repo while developing your UI?

**Reply by Festivejelly** [reply]
Okie dokie i'll have a crack I think I can make that work.  My preference is to have it over USB and the RPI will be the server. Whilst touch screens dont have tactile feedback having the various functions exposed to the web server means you can use any input device you like, be it a keyboard or touchscreen or something more custom like a custom keyboard or keypad.

Ill create the UI in my repo and ill fork this one i'll do a little video on it once im finished.

I'll have a think about how to handle the communication between the Nano and the RPI. The virtual button events might be a good starter just for proof of concept.
